data-model/src/main/java/org/hypertrace/core/datamodel/shared/TraceEventsGraph.java

```diff
@@ -0,0 +1,132 @@
+package org.hypertrace.core.datamodel.shared;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.hypertrace.core.datamodel.Edge;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.StructuredTrace;
+
+public class TraceEventsGraph {
+
+  /* there could be multiple roots for partial trace and incomplete instrumented app spans */
+  private final Map<ByteBuffer, List<Event>> parentToChildrenEvents;
+  private final Map<ByteBuffer, Event> childToParentEvents;
+
+
+  /* these containers should be unmodifiable after initialization as we're exposing them via getters */
+  private Map<ByteBuffer, Event> eventMap;
+  private Set<Event> rootEvents;
+
+  private TraceEventsGraph() {
+    this.childToParentEvents = new HashMap<>();
+    this.parentToChildrenEvents = new HashMap<>();
+  }
+
+  /**
+   * Create the instance with or without the full traversal graph includesFullGraph    Includes
+   * computing the full traversal graph or not
+   */
+  static TraceEventsGraph createGraph(StructuredTrace trace) {
+    TraceEventsGraph graph = new TraceEventsGraph();
+    graph.buildEventMap(trace);
```

Comment:
Both buildEventMap and buildRootEvents seem to be traversing through all the events.
We can save some time by moving calling buildParentChildRelationship first and then do both buildEventMap and buildRootEvents in one pass

Reply:
Done